### PR TITLE
rk3399 hdmi support (internal phy)

### PR DESCRIPTION
**Pull request by bukinr**
support tested on rk3399 only

### CODE REVIEW

File: bridges/dw_hdmi/aw_de2_dw_hdmi.c [hunk 1]
```diff
@@ -665,6 +922,34 @@ dw_hdmi_write(struct dw_hdmi_softc *sc, uint32_t reg, uint32_t val)
...
+dw_hdmi_register_ports(device_t dev)
+{
+	phandle_t ports, port;
+	phandle_t node, child;
+	char endp[DW_HDMI_MAX_PORTS];
+	int i;
+
+	node = ofw_bus_get_node(dev);
+
+	ports = ofw_bus_find_child(node, "ports");
```
Comment (evadot) [inline bridges/dw_hdmi/aw_de2_dw_hdmi.c:935]: This should use the ofw_graph functions.

File: bridges/dw_hdmi/aw_de2_dw_hdmi.c [hunk 2]
```diff
@@ -912,6 +1198,60 @@ MODULE_DEPEND(aw_de2_dw_hdmi, aw_de2_hdmi_phy, 1, 1, 1);
...
+rk_hdmi_clk_enable(device_t dev)
+{
+	struct rk_dw_hdmi_softc *sc;
+	uint64_t rate;
+	int error;
+	int i;
+
+	sc = device_get_softc(dev);
+
+	for (i = 0; i < RK_CLK_NENTRIES; i++) {
```
Comment (evadot) [inline bridges/dw_hdmi/aw_de2_dw_hdmi.c:1223]: The common clock will be enabled twice as we enable the required ones in dw_hdmi.

**Review by evadot — CHANGES_REQUESTED** [review]

File: bridges/dw_hdmi/aw_de2_dw_hdmi.c (hunk 1 above)
Comment (bukinr) [reply bridges/dw_hdmi/aw_de2_dw_hdmi.c]: Thanks. I removed this chunk as it was not needed

File: bridges/dw_hdmi/aw_de2_dw_hdmi.c (hunk 2 above)
Comment (bukinr) [reply bridges/dw_hdmi/aw_de2_dw_hdmi.c:1193]: Yes, but the rest of RK clocks have to be enabled *after* common clocks and *before* dw_attach(), so I have to enable them twice.

**Comment by evadot** [comment]
Merged

**Pull request closed without merging**